src/mongoc/mongoc-cluster.c

```diff
@@ -1214,19 +1214,31 @@ mongoc_cluster_fetch_stream (mongoc_cluster_t *cluster,
                              bson_error_t *error)
 {
    mongoc_topology_scanner_node_t *scanner_node;
+   mongoc_topology_t *topology;
    mongoc_cluster_node_t *cluster_node;
    mongoc_server_description_t *sd = NULL;
    mongoc_stream_t *stream = NULL;
+   int64_t timestamp;
+   bool single_threaded;
 
    bson_return_val_if_fail(cluster, NULL);
 
-   scanner_node = mongoc_topology_scanner_get_node(cluster->client->topology->scanner, server_id);
+   topology = cluster->client->topology;
+   mongoc_mutex_lock (&topology->mutex);
+
+   scanner_node = mongoc_topology_scanner_get_node(topology->scanner, server_id);
```

Comment:
You'll want to strip out the other uses of mutex_lock for the topology mutex as well.